### PR TITLE
[5.3] Unify join and builder syntax

### PR DESCRIPTION
**Pull request by acasar**
This is a massive cleanup of the join clause (~200 lines of production code removed). The idea is to extend the query builder and reuse its functionality instead of duplicating it. This has multiple benefits:
- no need to duplicate the logic for where clauses
- simplified join compilation (by reusing where clause compiler)
- unified syntax between join clause and query builder (i.e. "=" operator is now optional, access to additional clauses like `whereRaw`, `whereSub`, ...)

The only breaking change is the removal of `$where` boolean parameter in the `on` clause and the removal of some public `JoinClause` properties.

### CODE REVIEW

**Comment by willrowe** [comment]
Would there be any issues with using methods available on `Builder` that wouldn't necessarily be valid in a join statement? I'm thinking of methods like `select`, `orderBy`, or `groupBy`.

**Comment by acasar** [comment]
They will be ignored, same as with nested where... Currently nothing is stopping you from writing:

```php
User::where(function($query){
    $query->orderBy('foo');
})->get();
```
I first wanted to extract where builder in a separate class and use it for nested wheres & joins, but it would make merging 5.2 -> 5.3 a nightmare.

The only problem is if you introduce non-where bindings - the query would fail in both cases.

**Comment by srmklive** [comment]
Totally agree with @acasar in this regard. Seems a really good feature to have in 5.3

File: tests/Database/DatabaseEloquentBuilderTest.php [hunk 1]
```diff
@@ -506,7 +506,7 @@ public function testHasWithContraintsAndJoinAndHavingInSubquery()
         $builder = $model->where('bar', 'baz');
         $builder->whereHas('foo', function ($q) {
             $q->join('quuuux', function ($j) {
-               $j->on('quuuuux', '=', 'quuuuuux', 'and', true);
+               $j->where('quuuuux', '=', 'quuuuuux');
```
Comment (taylorotwell) [inline tests/Database/DatabaseEloquentBuilderTest.php]: Why did you need to change this test?

**Comment by taylorotwell** [comment]
Looks good. Thanks!

File: tests/Database/DatabaseEloquentBuilderTest.php (hunk 1 above)
Comment (acasar) [inline tests/Database/DatabaseEloquentBuilderTest.php]: The last param (true) was removed from the ``on`` clause since it no longer makes sense.
